### ISSUE TITLE
Optimization tracking issue

### ISSUE THREAD

**Issue by tomwhite** [tomwhite/cubed#339]
This is an umbrella issue for tracking the work on optimizations in Cubed.

## Creation optimizations
Making creation operations more efficient, typically by not materializing unnecessary data.
- [ ] #336

## Fusion optimizations
Currenly we fuse map blocks operations with one input, but there are more types of fusion we could implement.
- [ ] #337
  - This should be done early on, so changes to fuse or other DAG manipulation tasks don't need to be done twice
- [ ] #288
- [ ] Allow more control over optimizations applied
  - For example, allow users to switch on more aggressive optimizations
- [ ] #136
  - This change will fuse operations with multiple inputs
- [ ] #69
- [ ] Sibling fusion
  - This change will fuse operations that share the same inputs

## Reduction optimizations
Reduction operations like `sum` and `mean` could be improved by minimising the amount of data transferred.
- [ ] Support arbitrary block index functions in blockwise
  - Generalize blockwise to allow arbitrary mapping between input and output blocks to support `partial_reduce` and `tree_reduce`
- [ ] #331

## High-level query optimizations
Re-writing array expressions to an optimized form (before applying the optimizations above).
- [ ] #333

**Comment by TomNicholas** [comment]
Do any of the caveats on the [Scaling docs page](https://tom-e-white.com/cubed/user-guide/scaling.html) need updating after all these improvements? e.g. it currently says

> In theory multiple blockwise operations can be fused together, enhancing the performance further. However this has not yet been implemented in Cubed.

**Comment by tomwhite** [comment]
> Do any of the caveats on the [Scaling docs page](https://tom-e-white.com/cubed/user-guide/scaling.html) need updating after all these improvements? e.g. it currently says
> 
> > In theory multiple blockwise operations can be fused together, enhancing the performance further. However this has not yet been implemented in Cubed.

Yes, thanks for raising this. I've opened #381 to track documentation changes.
